servicetalk-grpc-utils/src/main/java/io/servicetalk/grpc/utils/LoggingGrpcLifecycleObserver.java

```diff
@@ -0,0 +1,198 @@
+/*
+ * Copyright © 2021 Apple Inc. and the ServiceTalk project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.servicetalk.grpc.utils;
+
+import io.servicetalk.buffer.api.Buffer;
+import io.servicetalk.grpc.api.GrpcLifecycleObserver;
+import io.servicetalk.grpc.api.GrpcStatus;
+import io.servicetalk.http.api.HttpHeaders;
+import io.servicetalk.http.api.HttpRequestMetaData;
+import io.servicetalk.http.api.HttpResponseMetaData;
+import io.servicetalk.logging.api.LogLevel;
+import io.servicetalk.logging.slf4j.internal.FixedLevelLogger;
+import io.servicetalk.transport.api.ConnectionInfo;
+
+import javax.annotation.Nullable;
+
+import static io.servicetalk.logging.slf4j.internal.Slf4jFixedLevelLoggers.newLogger;
+import static io.servicetalk.utils.internal.ThrowableUtils.combine;
+import static java.lang.System.nanoTime;
+import static java.util.concurrent.TimeUnit.NANOSECONDS;
+
+/**
+ * Logging implementation of {@link GrpcLifecycleObserver}.
+ */
+public final class LoggingGrpcLifecycleObserver implements GrpcLifecycleObserver {
+
+    private final FixedLevelLogger logger;
+
+    public LoggingGrpcLifecycleObserver(final String loggerName, final LogLevel logLevel) {
+        this.logger = newLogger(loggerName, logLevel);
+    }
+
+    @Override
+    public GrpcExchangeObserver onNewExchange() {
+        return new LoggingGrpcExchangeObserver(logger);
+    }
+
+    private static final class LoggingGrpcExchangeObserver implements GrpcExchangeObserver,
+                                                                      GrpcRequestObserver, GrpcResponseObserver {
+
+        private final long startTime = nanoTime();
```

Comment:
I didn't plan to test this class because there is not much to test except the produced logger formatting. More important that other tests verify callbacks are invoked as expected.